From c6d3434eb1eac02583c110a2f4f869218c7403fb Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 18:59:13 -0400
Subject: [fs] btrfs: fix a warning when iput a file

Message-id: <1407179109-13714-67-git-send-email-sandeen@redhat.com>
Patchwork-id: 86997
O-Subject: [RHEL7.1 PATCH 066/422] Btrfs: fix a warning when iput a file
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Wang Shilong <wangsl.fnst@cn.fujitsu.com>

commit 180589efde8a01b4a30af273f670ac81c8abf9c5
Author: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
Date:   Sat Dec 14 15:27:31 2013 +0800

    Btrfs: fix a warning when iput a file

    See the warning below:

    [ 1209.102076]  [<ffffffffa04721b9>] remove_extent_mapping+0x69/0x70 [btrfs]
    [ 1209.102084]  [<ffffffffa0466b06>] btrfs_evict_inode+0x96/0x4d0 [btrfs]
    [ 1209.102089]  [<ffffffff81073010>] ? wake_atomic_t_function+0x40/0x40
    [ 1209.102092]  [<ffffffff8118ab2e>] evict+0x9e/0x190
    [ 1209.102094]  [<ffffffff8118b313>] iput+0xf3/0x180
    [ 1209.102101]  [<ffffffffa0461fd1>] btrfs_run_delayed_iputs+0xb1/0xd0 [btrfs]
    [ 1209.102107]  [<ffffffffa045d358>] __btrfs_end_transaction+0x268/0x350 [btrfs]

    clear extent bit here to avoid triggering WARN_ON() in remove_extent_mapping()

    Signed-off-by: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index a954065..149e208 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -4509,6 +4509,8 @@ static void evict_inode_truncate_pages(struct inode *inode)
 
   node = rb_first(&map_tree->map);
   em = rb_entry(node, struct extent_map, rb_node);
+  clear_bit(EXTENT_FLAG_PINNED, &em->flags);
+  clear_bit(EXTENT_FLAG_LOGGING, &em->flags);
   remove_extent_mapping(map_tree, em);
   free_extent_map(em);
  }
-- 
1.7.1